File: SWEA/6808. 규영이와 인영이의 카드게임.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Solution
{
    public static void swap(ArrayList<Integer> cards, int x, int y)
    {
        int temp = cards.get(x);
        cards.set(x, cards.get(y));
        cards.set(y, temp);
    }

    public static void main(String[] args) throws Exception
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int tc = Integer.parseInt(br.readLine());
        for (int t = 1; t <= tc; t++)
        {
            sb.append("#").append(t).append(" ");
            //////////////////////////////////////
            boolean[] sel = new boolean[19]; // 1~18번까지 카드 선택 여부, true면 규영이꺼
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < 9; i++)
                sel[Integer.parseInt(st.nextToken())] = true;

            // 규영이가 내는 카드는 고정
            ArrayList<Integer> g_card = new ArrayList<>();
            ArrayList<Integer> i_card = new ArrayList<>();
            for (int i = 1; i <= 18; i++)
            {
                if(sel[i]) g_card.add(i);
                else i_card.add(i);
            }

            // 인영이 카드의 순열을 구해서 규영이와 대조하자.
            wincnt = 0;
            perm(g_card, i_card, 0);
            //////////////////////////////////////
            sb.append(wincnt).append(" ").append(fact9-wincnt).append("\n");
        }
        System.out.println(sb);
        
    }
    public static final int fact9 = 9*8*7*6*5*4*3*2;
    public static int wincnt = 0;
    public static void perm(ArrayList<Integer> g_card, ArrayList<Integer> i_card, int idx)
    {
        // 순열을 구한 상태
        if(idx == 9)
        {
            // 대소 비교를 통해서 규영이와 인영이의 합을 구한다.
            int g_sum = 0;
            int i_sum = 0;
            for (int i = 0; i < 9; i++)
            {
                int point = g_card.get(i) + i_card.get(i);
                if(g_card.get(i) > i_card.get(i))
                    g_sum += point;
                else
                    i_sum += point;
            }
            if(g_sum > i_sum)
                wincnt++;
            return;
        }

        // 인영이의 카드 순서를 바꿔보자
        for (int i = idx; i < 9; i++)
        {
            if(i != idx)
                swap(i_card, idx, i);
            perm(g_card, i_card, idx+1);
            swap(i_card, idx, i);
        }
    }
}

File: SWEA/6808.규영이와 인영이의 카드게임.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <cstdio>
using namespace std;

int  g_cards[9];
int  i_cards[9];
int wincnt;
int fact9 = 362880;

void swap(int * arr, int x, int y)
{
	int temp = arr[x];
	arr[x] = arr[y];
	arr[y] = temp;
}

void perm(int n)
{
	if (n == 9)
	{
		// 점수의 합을 구하기 위해서 규형이와 인영이 점수를 누적하자
		int g_score = 0;
		int i_score = 0;
		for (int i = 0; i < 9; i++)
		{
			// 규영이가 이겼으면
			if (g_cards[i] > i_cards[i])
			{
				g_score += g_cards[i] + i_cards[i];
			}
			else
			{
				i_score += g_cards[i] + i_cards[i];
			}
		}
		if (g_score > i_score)
		{
			wincnt++;
		}
		return;
	}
	for (int i = n; i < 9; i++)
	{
		if(i != n)
			swap(i_cards, i, n);
		perm(n + 1);
		swap(i_cards, i, n);
	}
}

int main()
{
	size_t T = 0;
	scanf("%u", &T);
	for (size_t i = 1; i <= T; i++)
	{
		bool selected[19];
		for (size_t j = 0; j < 19; j++)
			selected[j] = false;

		for (size_t j = 0; j < 9; j++)
		{
			int temp;
			scanf("%d", &temp);
			g_cards[j] = temp;
			selected[g_cards[j]] = true;
		}

		int idx = 0;
		for (size_t j = 1; j < 19; j++)
		{
			if (selected[j] == false)
			{
				i_cards[idx++] = j;
			}
		}
		wincnt = 0;
		perm(0);
		cout << '#' << i << ' ' << wincnt << ' ' << (fact9 - wincnt) << '\n';
	}
    return 0;
}